### PR TITLE
dbt-materialize: allow additional `name` param in index config

### LINKED ISSUE #11967: dbt-materialize: index config should allow optional `name` param
### What version of `dbt-materialize` are you using?

`Unreleased`

### How did you install `dbt-materialize`?

Built from source

### What was the issue?

In #11538, we deprecated the custom index materialization and made index creation part of the model definition as a config. The macro follows the implementation of [`postgres__get_create_index_sql`](https://github.com/dbt-labs/dbt-core/blob/22416980d19888d7bbbcfcda639c16d37a80b948/plugins/postgres/dbt/include/postgres/macros/adapters.sql#L17-L30), where:

> For the index’s name, dbt uses a hash of its properties and the current timestamp, in order to guarantee uniqueness and avoid namespace conflict with other indexes.

This means that **users can no longer specify a custom index name**, and are instead stuck with the auto-generated name (see `104987`): 

image (6)

We should consider providing an optional `name` param to match the expected behaviour.

<hr>

🤟 Thanks to @markns for reporting this!

### PR DESCRIPTION
**Pull request by morsapaes**
### Motivation

  * This PR adds a known-desirable feature.

    Fixes #11967.

### Tips for reviewer

I'm not happy with the macro-side implementation. My first thought was to use a [default filter](https://support.sendwithus.com/jinja/default/), like:

```
{%- set index_name = index_config.name | default(index_config.render(relation)) -%}
```

, but it was tripping up with `None`. Open to suggestions on how to make this look less gross, @benesch. 👹 

### Testing

- [x] This PR has adequate test coverage / QA involvement has been duly considered.
   
   Modified the `models/test_materialized_view_index.sql` test in `materialize.dbtspec`.

#### Create index with `name`

```
{{ config(materialized='materializedview',
          indexes=[{'name': 'mv_index', 'columns': ['symbol']}]) }}
```

```sql
materialize=> show indexes on avg_bid;
 cluster | on_name |      key_name       | seq_in_index | column_name | expression | nullable | enabled
---------+---------+---------------------+--------------+-------------+------------+----------+---------
 default | avg_bid | avg_bid_primary_idx |            1 | symbol      |            | t        | t
 default | avg_bid | mv_index            |            1 | symbol      |            | t        | t
(2 rows)
```

#### Create index without `name`

```
{{ config(materialized='materializedview',
          indexes=[{'columns': ['symbol']}]) }}
```

```sql
materialize=> show indexes on avg_bid;
 cluster | on_name |             key_name             | seq_in_index | column_name | expression | nullable | enabled
---------+---------+----------------------------------+--------------+-------------+------------+----------+---------
 default | avg_bid | avg_bid_primary_idx              |            1 | symbol      |            | t        | t
 default | avg_bid | c57c03b083ba7092774eb88939d310cb |            1 | symbol      |            | t        | t
(2 rows)
```

<hr>

Once this is settled and merged, I'll tackle the open documentation tasks mentioned in #11538.

### CODE REVIEW

File: misc/dbt-materialize/MAINTAINERS.md [hunk 1]
```diff
@@ -29,28 +29,31 @@
...
 
    ```
-   brew install materialize/materialize/materialized
+   brew install MaterializeInc/materialize/materialized
+
    materialized
    ```
 
-6. Run tests:
+6. Install `pytest-dbt-adapter` and run tests:
```
Comment (morsapaes) [inline misc/dbt-materialize/MAINTAINERS.md:37]: Are these instructions a bit useless now that you need a running Kafka service to run the tests?

**Comment by morsapaes** [comment]
Ah, sounds like releasing should be it's own PR, so removing the version bump from this one! I'll create a release PR once this is merged.

**Comment by morsapaes** [comment]
One scenario I forgot to test before but now see **fails** is using the same index name in two different models. In this case, the index is created for the first model, but silently ignored for the second one:

**`avg_bid.sql`**
```
{{ config(materialized='materializedview',
          indexes=[{'name': 'mv_index', 'columns': ['symbol']}]) }}
```

**`market_orders.sql`**
```
{{ config(materialized='view',
          indexes=[{'name': 'mv_index', 'columns': ['bid_price']}]) }}
```

**`dbt run`**

```
11:16:40  1 of 3 START source model public.market_orders_raw.............................. [RUN]
11:16:40  1 of 3 OK created source model public.market_orders_raw......................... [CREATE SOURCE in 0.02s]
11:16:40  2 of 3 START view model public.market_orders.................................... [RUN]
11:16:40  2 of 3 OK created view model public.market_orders............................... [CREATE VIEW in 0.02s]
11:16:40  3 of 3 START materializedview model public.avg_bid.............................. [RUN]
11:16:40  3 of 3 OK created materializedview model public.avg_bid......................... [CREATE VIEW in 0.01s]
```

**`materialized`**

```
materialize=> SHOW INDEXES ON market_orders;
 cluster |    on_name    | key_name | seq_in_index | column_name | expression | nullable | enabled
---------+---------------+----------+--------------+-------------+------------+----------+---------
 default | market_orders | mv_index |            1 | bid_price   |            | t        | t
(1 row)

materialize=> SHOW INDEXES ON avg_bid;
 cluster | on_name |      key_name       | seq_in_index | column_name | expression | nullable | enabled
---------+---------+---------------------+--------------+-------------+------------+----------+---------
 default | avg_bid | avg_bid_primary_idx |            1 | symbol      |            | t        | t
(1 row)
```

**Comment by benesch** [comment]
> Is there a way to surface ERROR: catalog item 'mv_index' already exists all the way to dbt and make the run fail? I half-assumed this would just...happen.

I ... would have expected that too. Is there possibly a bug in dbt?

Also, question: it seems like this would be useful for PostgreSQL indexes in dbt too. Have we reached out to ask if dbt-postgres in interested in this feature? If yes, then we could plausibly PR this against dbt-postgres and not be responsible for the code long term. If no, would be good to know if something goes wrong in dbt if you let users specify their own names for indexes.

**Comment by morsapaes** [comment]
> If no, would be good to know if something goes wrong in dbt if you let users specify their own names for indexes.

Ah, I forgot that I'd bumped into [this issue](https://github.com/dbt-labs/dbt-core/issues/1945#issuecomment-576714925) during "recon"; forcing a random unique name was an [implementation decision](https://github.com/dbt-labs/dbt-core/pull/3106) to work around the fact that dbt seems unable to handle name collisions in indexes properly.

Since cleanly allowing an index `name` sounds like a bigger discussion, I'll go ahead and cut a `dbt-materialize` release with the default implementation to at least enable users waiting for this to make the switch (and avoid existing index-related issues like #11969), and kick off a thread in the dbt community.

**Comment by benesch** [comment]
🎉 sounds good! Btw, I'd be curious to follow along the thread in the dbt community, if you've got a link!

**Review by dataders — APPROVED** [review]
love to see that this is being implemented! I'm having trouble disambiguating what's different from this PR compared to #11613, but I get the general gist. Small suggestions, but looks good to me! Glad to see Postgres's implementation be picked up by so many other dbs!

**Review by dataders — APPROVED** [review]
love to see that this is being implemented! I'm having trouble disambiguating what's different from this PR compared to #11613, but I get the general gist. Small suggestions, but looks good to me! Glad to see Postgres's implementation be picked up by so many other dbs!

File: misc/dbt-materialize/MAINTAINERS.md (hunk 1 above)
Comment (morsapaes) [reply misc/dbt-materialize/MAINTAINERS.md:37]: Thanks for the reminder, @dataders! We have this queued as a follow-up task from #12141!

**Comment by morsapaes** [comment]
Thanks for checking this out, @dataders! The difference between this PR and #11613 is that it diverges (or, tries to) from the base Postgres implementation, which doesn't allow users to specify a `name` for the index and instead forces an auto-generated name. This has been brought up by one of our users as not-so-nice, so we're exploring our options! Ideally, we'd find a solution that also works for `dbt-postgres` and that we can contribute upstream.

I'll check your suggestion IWRT Firebolt, and maybe bring it up for discussion in the dbt community Slack. We're trying to keep the `dbt-materialize` adapter as close to the implementation of `dbt-postgres` as possible.

File: misc/dbt-materialize/dbt/adapters/materialize/impl.py [hunk 2]
```diff
@@ -30,3 +58,7 @@ def _link_cached_relations(self, manifest):
         #
         # [0]: https://github.com/dbt-labs/dbt-core/blob/13b18654f03d92eab3f5a9113e526a2a844f145d/plugins/postgres/dbt/adapters/postgres/impl.py#L126-L133
         pass
+
+    @available
+    def parse_index(self, raw_index: Any) -> Optional[MaterializeIndexConfig]:
+        return MaterializeIndexConfig.parse(raw_index)
```
[the remark this answers is not in the document]
Comment (morsapaes) [reply misc/dbt-materialize/dbt/adapters/materialize/impl.py:65]: Really appreciate the pointer! I'll have a deeper look later this week!

**Comment by dataders** [comment]
> We're trying to keep the `dbt-materialize` adapter as close to the implementation of `dbt-postgres` as possible.

totally understandable, though the addition of index creation for PostgreSQL is rather new (https://github.com/dbt-labs/dbt-core/pull/3106), so there's no established canon yet really. If there's a growing consensus about how indexes should be created across different dbs, we can certainly modify the default implementation! Just something to think about.

**Comment by morsapaes** [comment]
Closing as this was implemented in #14195. We should double-check whether the naming collision issue still manifests there, @ahelium!

**Pull request closed without merging**